### DISCUSSION TITLE
Retrieve all modules a physical node is assigned to in python API

### DISCUSSION THREAD

**Discussion by u-xixi** [mapequation/infomap#272] [Q&A] [answered]
Hi

Sorry if i'm spamming you again. I'm using the python API to partition a multilayer network. And I want to get _all_ the modules each physical node is assigned to, using the Infomap object.  I know I can write out a `.clu` file and parse that, but it seems like a detour.

I know that I can call `get_modules(states=False)` to retrieve node_id: module_id assignment. But in this case, if the physical node has states in multiple modules, I can get only one of them.  `get_modules(states=False)` instead has state_id as the key. But I didn't find out how to bridge the state_id and their physical node_id directly.

Maybe a bit more context: I study bioinformatics and the nodes are genome fragments. I'm exploring if the overlapping clustering could even perform better than the traditionally non-overlapping way. My network was created layer by layer, in each layer, link by link, using `add_multilayer_intra_link()`. The state nodes are made internally by Infomap, so I would not know what states each physical node has.

thanks in advance,
Xixi

**Comment by danieledler** [comment, marked as answer]
Hi @u-xixi,

For a multilayer network, you can access the `node_id` and `layer_id` together with `module_id` when you iterate the (state) nodes:

```py
for node in im.nodes:
    print(node.node_id, node.layer_id, node.module_id)
```

This will give you overlapping modules, as the same `node_id` may occur multiple times for different `layer_id`.

See [this example](https://github.com/mapequation/infomap/blob/master/examples/python/multilayer.py).
